Allow Background Subtraction to be Saved and Loaded from Batch Files

cailafinn opened the pull request:
**Description of work**

**Purpose of work**
The original implementation of this subtraction method allowed the values for the background workspace and the scale factor to be set in an `Options` column. This column's cells are comma-separated lists of parameters. This is incompatible with the `.csv` format used for the batch files by the SANS group, and so the values in the options column are not stored in the batch files. 

This new background subtraction feature is reported to eventually be in high enough use that this information needs to be stored in the batch so that it can be edited alongside the other parameters present on the table. 

**Report to:** Robert Dalgliesh

**Summary of work**
Creates two new columns on the Runs tab of the ISIS SANS Interface for the "Background Workspace" and the "Scale Factor". In order to avoid cluttering an already busy interface, the columns do not appear on the table until the "Scaled Background Subtraction" checkbox is checked.

**Further detail of work**
- Removed the background workspace and scale factor parameters from the options box
- Allowed the new bg subtraction parameters to be saved and loaded into and out of the batch files (`.csv` format)
- Added validation when loading a batch file that both the scale factor and background workspace parameters are set (both are required for the reduction, and the batch files are often edited in excel)


**To test:**
You may need to change the value of the `.toml` user file for the reduction mode from "All" to "Merged". This is called `MaskFile.toml` in the training course data. You should confirm this 

1. Boot up workbench and the ISIS SANS interface. 
2. Load the user file and the batch file from the LOQ Demo in the TrainingCourseData
4. Process All (this will produce your "background" workspaces, something like `first_time_merged_1D_2.2_10.0` or `first_time_HAB_1D_2.2_10.0`)
5. Check the `Scaled Background Subtraction` checkbox
6. Copy and paste a new row and set `Background Workspace` to the name of the workspace you produced in step 3, and `Scale Factor` to 0.9.
7. Process this new row. 
8. Check the two output workspaces you should now have. If it has worked properly, the subtracted one's values should be 10% of the unsubtracted one's. 
9. Check that the batch can be saved and loaded and the background workspace and scale factor cells are preserved between loads. 

Fixes #35958


*This does not require release notes* because **it makes changes to a new feature for this release.**




---

#### Reviewer ####

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

#### Gatekeeper ####

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

Commit:
Add background sub params to RowEntries

RE #35958

Commit:
Allow parsing of background subtraction parameters

RE #35958

Commit:
Move bgsub params out of options col parsing

RE #35958

Commit:
Add bgsub parameters to GUI table

We don't want these to clutter up the GUI all of the time, so we keep
them hidden behind a checkbox that enables them.

RE #35958

Commit:
Update documentation & testing instructions

RE #35958

Commit:
Amend test after addition of bgsub params

RE #35958

rbauststfc commented on dev-docs/source/Testing/SANSGUI/ISISSANSGUITests.rst:
```diff
@@ -171,8 +171,9 @@ Processing
+#. Check the ``Scaled Background Subtraction`` checkbox.
+#. In the ``BackgroundWorkspace`` column, enter the name of the merged workspace you took note of before.
+#. In the ``ScaleFactor`` column, enter ``0.9``.
 #. Select this new row and click ``Process Selected``.
```
I remembered when testing that there is a known behaviour where changing information in the batch table causes manually changed GUI settings to revert to the user file value. Is there something earlier in these test instructions that means this won't be an issue here, or do we need to set the reduction mode only after we've finished editing the batch table?

ConorMFinn replied:
Good point! I'd forgotten about that feature. I'll add that in

rbauststfc replied:
Thinking about it, you'll probably need it both at the beginning and later on unfortunately because it will need to be set to merged before running the first reduction to create the background workspace. Then, after entering the background subtraction settings, it will change back to all so it will need to be changed again. Perhaps one approach could be to set everything up in the batch table, switch to merged reduction and then use the button that processes all the rows in sequence rather than doing them one at a time? We should be able to provide the name that will be output for the merged background workspace in the testing instructions.

ConorMFinn replied:
I think I'm going to alter the instructions to just change the user file. This is the last step of the runs tab testing, so it should be safe to change it at this point without destroying all the future tests.

rbauststfc replied:
I think the only consideration is if people change it then we would need to make sure it's correct next time they start the manual testing instructions again, in case they haven't refreshed their training data (I know that I don't re-download it very frequently).

rbauststfc commented on docs/source/interfaces/isis_sans/Runs Tab.rst:
```diff
@@ -109,17 +109,26 @@ Table Columns
+Additional hidden columns can be shown on the table by checking the ``Scaled Background Subtraction`` checkbox above
+the table on the right hand side.
+
+By setting **both** ``Background Workspace`` and ``Scale Factor`` a Scaled Background Subtraction will be
```
Is it worth making a note here about the fact that if the can values are in the row then we do sample minus can minus scaled background? When we were testing with the scientists, it sounded like this might be a good thing to highlight.

rbauststfc reviewed (commented):
The changes are all working really well. I was able to save and load batch files with no issue and it was fully compatible with existing batch files that don't have information for the new columns. Settings entered into the batch table were saved out and loaded back in correctly. The added columns show and hide correctly and are easy to use. The background subtraction is completing accurately, showing that the column values are being passed through to the reduction correctly. The code changes all look good.

I had just a couple of minor comments about docs, but otherwise this looks good, and I know the scientists will be very grateful to have this change in the release!

Commit:
Amend docs to clarify confusion from testing

There was some confusion over what the additional background subtraction
did as supplying the Can information is also a form of background
subtraction.

Since the background workspace can in theory contain anything as long as
the workspace dimensions are the same, it's best to be explicit about
what it's doing, so some examples have been added to illustrate this.

RE #35958

Commit:
Use User file in the manual testing

The settings get overwritten when changing the runs table. So this
method is more stable

RE #35958

rbauststfc reviewed (approved):
All requested documentation changes have now been resolved (thanks Conor!), so this is ready to be merged in.
